src/parser/jobs/ast/modules/ArcanaTracking.tsx

```diff
@@ -0,0 +1,509 @@
+	SPEAR = STATUSES.SPEAR_DRAWN.id,
+	EWER = STATUSES.EWER_DRAWN.id,
+	SPIRE = STATUSES.SPIRE_DRAWN.id,
+	LORD_OF_CROWNS = STATUSES.LORD_OF_CROWNS_DRAWN.id,
+	LADY_OF_CROWNS = STATUSES.LADY_OF_CROWNS_DRAWN.id,
+}
+
+export enum SleeveType {
+	NOTHING = 0,
+	ONE_STACK = 1,
+	TWO_STACK = 2,
+}
+
+interface PullEvent extends Event {
+	type: 'pull'
+}
+
+export interface CardState {
+	lastEvent: CastEvent | PullEvent | DeathEvent
+	drawState: DrawnType
+	sealState: SealType[]
+	sleeveState: SleeveType
+}
+
+// TODO: Try to track for when a seal was not given on pull due to latency?
+export default class ArcanaTracking extends Module {
+	static handle = 'arcanaTracking'
+	static dependencies = [
+		'precastAction', // eslint-disable-line @xivanalysis/no-unused-dependencies
+		'precastStatus', // eslint-disable-line @xivanalysis/no-unused-dependencies
+	]
+	static title = t('ast.arcana-tracking.title')`Arcana Tracking`
+	static displayOrder = DISPLAY_ORDER.ARCANA_TRACKING
+
+	private cardStateLog: CardState[] = [{
+		lastEvent: {
+			type: 'pull',
+			timestamp: this.parser.fight.start_time,
+			targetIsFriendly: true,
+			sourceIsFriendly: true,
+		},
+		drawState: DrawnType.NOTHING,
+		sealState: CLEAN_SEAL_STATE,
+		sleeveState: SleeveType.NOTHING,
+	}]
+	private pullStateInitialized = false
+	private pullIndex = 0
+
+	private prepullArcanas: BuffEvent[] = []
+
+	protected init() {
+		this.addHook('cast', {abilityId: CARD_ACTIONS, by: 'player'}, this.onCast)
+
+		this.addHook('applybuff', {abilityId: ARCANA_STATUSES, by: 'player'}, this.onPrepullArcana)
+		this.addHook('removebuff', {abilityId: ARCANA_STATUSES, by: 'player'}, this.offPrepullArcana)
+
+		this.addHook('removebuff', {abilityId: DRAWN_ARCANA, by: 'player'}, this.offDrawnStatus)
+		this.addHook('death', {to: 'player'}, this.onDeath)
+		this.addHook('complete', this.onComplete)
+	}
+
+	normalise(events: any) {
+		const startTime = this.parser.fight.start_time
+		let prepullSleeve = false
+		const sleeveDrawLog: CastEvent[] = []
+
+		for (const event of events) {
+			if ((event.timestamp > startTime && !prepullSleeve)
+				|| event.timestamp - startTime >= (STATUSES.SLEEVE_DRAW.duration * 1000)
+				) {
+				// End loop if: 1. Max duration of sleeve draw status passed 2. No prepull sleeve detected
+				break
+
+				} else if (!prepullSleeve && event.type === 'applybuff' && STATUSES.SLEEVE_DRAW.id === event.ability.guid
+				&& event.timestamp <= startTime) {
+					prepullSleeve = true
+				} else if (prepullSleeve && event.type === 'cast' && ACTIONS.DRAW.id === event.ability.guid) {
+					sleeveDrawLog.push(event)
+				} else if (prepullSleeve && event.type === 'removebuff' && STATUSES.SLEEVE_DRAW.id === event.ability.guid) {
+					prepullSleeve = false
+			} else {
+				continue
+			}
+		}
+
+		if (sleeveDrawLog.length > 0 && sleeveDrawLog.length <= 2) {
+			this.cardStateLog[0].sleeveState = sleeveDrawLog.length
+		}
+
+		return events
+
+	}
+
+	public get getCardLogs() {
```

Comment:
not sure i'm following what's going on here - `cardStateLog` should be `[]` if `length === 0` anyway, and it's initiated with an initial entry.

---

also avoid naming getters with `get...`, it looks kinda weird at call sites

Reply:
Okay, this one's my bad - this is leftover code from the 4.0 impl of this module, which did have a use for that check. I'll just make it a simple return getter.